### ISSUE TITLE
HTTPPost example is not working with vodafone sim

### ISSUE THREAD

**Issue by radar155** [blemasle/arduino-sim808#2]
I've just opened the example, changed pins and apn configuration, enabled debug.
This is the output from serial monitor
```
N: Powering on SIM808...
Init...
Waiting for echo...

-->AT
Waiting for echo...

-->AT
<--�Waiting for echo...

-->AT
<--AT
<--OK
<--
<--RDY

-->ATE0
<--
<--+CFUN: 1
<--
<--+CPIN: READY
<--ATE0
<--OK

-->AT+CGREG?
<--
<--+CGREG: 0,2
<--
<--OK

-->AT+CSQ
<--
<--+CSQ: 0,0
<--
<--OK
N: No network yet...

-->AT+CGREG?
<--
<--Call Ready
<--
<--SMS Ready
<--
<--+CGREG: 0,1
<--
<--OK

-->AT+CSQ
<--
<--+CSQ: 16,0
<--
<--OK
N: Network is ready.
N: Attenuation : -82 dBm, Estimated quality : 16
N: Powering on SIM808's GPRS...

-->AT+CIPSHUT
<--
<--SHUT OK

-->AT+CGATT=1
<--
<--OK

-->AT+CSTT="mobile.vodafone.it","",""
<--
<--OK

-->AT+SAPBR=1,1
<--
<--OK
N: Sending HTTP request...

-->AT+HTTPTERM
<--
<--ERROR

-->AT+HTTPINIT
<--
<--OK

-->AT+HTTPPARA="REDIR","1"
<--
<--OK

-->AT+HTTPPARA="CID","1"
<--
<--OK

-->AT+HTTPPARA="URL","http://httpbin.org/anything"
<--
<--OK

-->AT+HTTPPARA="CONTENT","text/plain"
<--
<--OK

-->AT+HTTPDATA=16,10000
<--
<--DOWNLOAD

-->This is the body<--
<--OK

-->AT+HTTPACTION=1
<--
<--OK
<--
<--+HTTPACTION: 1,601,0

-->AT+HTTPREAD=0,0
<--
<--ERROR
N: Server responsed : 601
N: This is the body
```

As you can see, in this example i didn't even changed the url, but it's not working with any url. HttpGet give the same error. Apn configuration is ok, my simcard is working and I can use with success http request by using [this](https://github.com/MarcoMartines/GSM-GPRS-GPS-Shield) library. It handles both GPRS Setup and http request with different AT COMMANDS than those you used. For HTTP he is using TCP AT Commands, for GPRS Setup he is using a slightly different AT command flow and parameters (but i'm not an expert). I can't understand if the problem is with HTTP Commands or with GPRS Setup.

**Comment by radar155** [comment]
I think that using the TCP methods would be overcomplicated. Simcom has released HTTP AT Commands officially (that you are already logically using).
I citied the TCP stack used by the other library just to say, but even if now something isn't working, I guess using HTTP stack is cleaner.

**Comment by blemasle** [comment]
The HTTP stack is very limited : you can't add custom header, and HTTPS is unreliable (7 requests out of 10 fail on my unit).

Anyway, I quickly found that CONTYPE GPRS was indeed missing in the code, and it worked fine on my unit because the bearer profile was never reset (which could mean that the network power off is not doing all of its jobs either btw). After resetting the bearer profile, I could reproduced your issue straight away.

Could you test to checkout the repo on the pull request I just opened and tell me if it fixes your issue. It certainly fixes it on my unit. See #3

**Comment by radar155** [comment]
Unfortunately I have the same error.
```
Init...
Waiting for echo...

-->AT
Waiting for echo...

-->AT
<--�Waiting for echo...

-->AT
<--AT
<--OK
<--
<--RDY

-->ATE0
<--
<--+CFUN: 1
<--
<--+CPIN: READY
<--ATE0
<--OK

-->AT+CGREG?
<--
<--+CGREG: 0,2
<--
<--OK

-->AT+CSQ
<--
<--+CSQ: 0,0
<--
<--OK
N: No network yet...

-->AT+CGREG?
<--
<--Call Ready
<--
<--SMS Ready
<--
<--+CGREG: 0,1
<--
<--OK

-->AT+CSQ
<--
<--+CSQ: 17,0
<--
<--OK
N: Network is ready.
N: Attenuation : -80 dBm, Estimated quality : 17
N: Powering on SIM808's GPRS...

-->AT+CIPSHUT
<--
<--SHUT OK

-->AT+CGATT=1
<--
<--OK

-->AT+SAPBR=0,1
<--
<--ERROR

-->AT+SAPBR=3,1,"CONTYPE","GPRS"
<--
<--OK

-->AT+CSTT="mobile.vodafone.it","",""
<--
<--OK

-->AT+SAPBR=1,1
<--
<--OK
N: Sending HTTP request...

-->AT+HTTPTERM
<--
<--ERROR

-->AT+HTTPINIT
<--
<--OK

-->AT+HTTPPARA="REDIR","1"
<--
<--OK

-->AT+HTTPPARA="CID","1"
<--
<--OK

-->AT+HTTPPARA="URL","http://httpbin.org/anything"
<--
<--OK

-->AT+HTTPPARA="CONTENT","text/plain"
<--
<--OK

-->AT+HTTPDATA=16,10000
<--
<--DOWNLOAD

-->This is the body<--
<--OK

-->AT+HTTPACTION=1
<--
<--OK
<--
<--+HTTPACTION: 1,601,0

-->AT+HTTPREAD=0,0
<--
<--ERROR
N: Server responsed : 601
N: This is the body
```

**Comment by blemasle** [comment]
Sorry to hear that :(
A few things I can think of right now
 * Is the network led flashing like three times a second after that ? If not, then the GPRS is not connected.
 * What is the return of the `ATI` command ? This is to be sure that firmware is up to date as old firmwares [do not seem to include the HTTP stack](https://www.raviyp.com/embedded/203-sim900-sim800-not-working-possible-reasons)
 * A few weeks ago, I [removed the `AT+CIICR`](https://github.com/blemasle/arduino-sim808/commit/fe5e336201c30b0bfd39e4fcb5c43c8fc0d2040c#diff-d2dbbfb21ebd8cb30e4790cde2cb8476L54) command call in the GPRS initialization sequence as it seemed to have no effect in that use case. Maybe you could try to add it back and see how it goes ?
 * Maybe you could try to reset the default config by sending `ATZ` and `AT&F` just in case other librairies have left a setting that's messing with this one. Use that in last resort though as after that, It won't be possible to find what's causing the issue easily.

You can sent manual commands easily by uploading the `Tester.ino` sketch, then sending `send at Z` or `send at &F` for instance. Sadly I can't check this on my side at the moment, as I will be on the move for the next couple of days. I'll try to come up with other ideas and post them here though.

Keep me posted !

**Comment by radar155** [comment]
- Yes, after that the newtork led is flashing like three times per second.

- Return value of ATI command: `SIM808 R14.18`

- I reintroduced the AT+CIICR command call in the GPRS initialization sequence, just after the AT+SAPBR=1,1 command call. AT+CIICR returns OK, but the 601 error is still present.

- I sent ATZ and AT&F commands. Both returns OK, but nothing changes. About this point i have a question: what is the RESET you are sending in the `SIM808.Power.cpp` module using RST Pin doing? I thought it was like an hard reset that not only restarts the sim808 module, but also reset all settings to default values (shutting down also gps and gprs antennas). Does sim808 has some persistent memory to remember some settings?

Did you see on my second log that the `AT+SAPBR=0,1` command is returning ERROR? I don't know if it can be usefull.

**Comment by radar155** [comment]
Here we are.
I just replaced the `+CSTT=APN,user,password` with `+SAPBR=3,1,APN`, according to [this](https://www.raviyp.com/embedded/194-sim900-gprs-http-at-commands) tutorial. Also, don't need to use the `+CIICR` command.
Now HTTP Post and Get are properly working!

**Comment by blemasle** [comment]
Glad to hear that, and sorry for the inconvenience :)
It's sad `+CSTT` cannot be used as it was a way to save some memory space.

The SIM808 does have some persistent memory and many things can and are stored in the demodule.

It is a hard reset, but it actually like a reset on your computer, just rebooting he thing. That's the easiest and the safest way to get the device to reboot in every situation.

As soon as I can, I'll try to update and publish a new version of the library with all those changes while trying to keep size down. Until then, keep me posted with every bug you find ! :p as I said in the README, I made the library for my personal use at first and while I used it extensively for the past year, I'm sure something will come up with someone else use case...

**Comment by blemasle** [comment]
Yep about the `AT+SAPBR=0,1` returning `ERROR`. I'm pretty sure that's why it worked after using `AT+SAPBR=3,1`. The first is supposed to reset the profile `1` while the second configure it. So I guess something was wrong with the reset.

I'll try to play a bit more with the reset of the profile to make it work in all cases. It always work on disabling the GPRS on my board but on enabling the GPRS that's another story...

I'll have to play with `CSTT` to see if it does set the bearer profile parameters or not. The documenting is quite unclear about what commands do in general, and I'm no AT commands expert either :)

**Comment by blemasle** [comment]
Hi @radar155 , and sorry for the delay (moving to another country, no sim with data etc).
I reverted most of the changes I made a few months ago using `+CSTT` instead of `+SAPBR`. It turns out that `CSTT` does not set the bearer profile and it was indeed only working on my unit because the bearer profile was previously set.

So now it should be working in all cases, but the `SAPBR=0,1` was still returning `ERROR` no matter where I put it in the sequence so I removed it for now and left it in the power off sequence. That will do the trick for now. I'll try to make the bearer profile reset to work in all cases later (the thing is resetting parameters that are already there could be mandatory for it to work).

Let me now if this works for you, and if so, I'll published the library with those changes.
